packages/components/src/sites-table/use-sites-table-filtering.tsx

```diff
@@ -0,0 +1,81 @@
+import { useFuzzySearch } from '@automattic/search';
+import { useI18n } from '@wordpress/react-i18n';
+import { useMemo } from 'react';
+import { SitesCountBadge } from './sites-count-badge';
+import type { Tab } from '../tab-panel';
+// eslint-disable-next-line no-restricted-imports
+import type { SiteExcerptData } from 'calypso/data/sites/site-excerpt-types';
+
+interface SitesTableFilterOptions {
+	status?: string;
+	search?: string;
+}
+
+interface UseSitesTableFilteringResult {
+	filteredSites: SiteExcerptData[];
+	tabs: Tab[];
+	selectedTabHasSites: boolean;
+}
+
+export function useSitesTableFiltering(
+	allSites: SiteExcerptData[],
+	{ status = 'all', search }: SitesTableFilterOptions
+): UseSitesTableFilteringResult {
+	const { __ } = useI18n();
+
+	const [ tabs, filteredByStatus ] = useMemo( () => {
+		const tabs: Tab[] = [
+			{ name: 'all', title: __( 'All' ) },
+			{ name: 'launched', title: __( 'Launched' ) },
+			{ name: 'private', title: __( 'Private' ) },
+			{ name: 'coming-soon', title: __( 'Coming soon' ) },
```

Comment:
For now, this is fine. However, if this would be truly reusable, we would ask for the tabs (and statuses) as parameters. Do you have any idea on how to do this, keeping in mind that `filterSites` would also need to be possibly redefined by the user?

Reply:
> keeping in mind that `filterSites` would also need to be possibly redefined by the user?

I think this is the reason why I like the site categories and filtering logic being defined in the hook. The filtering logic seems like something someone else would want to use (since it's a bit tricky to figure out what counts as what type of site), and since the categories and filtering logic are coupled (you can't define the switch statement without knowing which categories exist) I think it's good to bundle them together.

Some examples of how it could be re-used:

```js
// A component that just wants to show a table of launched sites
function LaunchedSitesTable() {
	const { data = [] } = useSiteExcerptsQuery();

	const { filteredSites } = useSitesTableFiltering( data, { status: 'launched' } );

	return <SitesTable sites={ filteredSites } />
}

// A component that searches a table but has no tabs
function SearchableSitesTable() {
	const { data = [] } = useSiteExcerptsQuery();
	const [ search, setSearch ] = useState( '' );

	const { filteredSites } = useSitesTableFiltering( data, { search } );

	return (
		<div>
			<input value={ search } onChange={ setSearch } />
			<SitesTable sites={ filteredSites } />
		</div>
	);
}
```

By adding this particular type of filtering (filtering by launch status) to `@automattic/components` it is privileging it over other types of filtering. Why shouldn't filtering by Plan be just as important? But I don't see why we couldn't have a shared `useSitesTablePlanFiltering()` hook too.

Reply:
Sorry, I mean being able to pass the tabs and filter predicates as parameters. Something like:

```ts
useSitesTableFiltering( data, {
  tabs: [
    { name: 'all', title: __( 'All' ), predicate: (site) => site },
    { name: 'private', title: __( 'Private' ), predicate: (site) => site.is_private }, 
  ]
} )
```

This way the user could determine custom filters as they like. We could have our current four tabs as the default argument.

Reply:
@p-jackson any thoughts on this one?

Reply:
@zaguiini I _do_ like it!

I still think it's ok to put our "launch status" logic inside the `packages/components/src/sites-table/` folder because I think it's reasonable to expect other users of the sites table would want to use the same predicates. But I think `useSitesTableFiltering` could be the building block for multiple hooks that we export:
- `packages/components/src/sites-table/use-sites-table-launch-filtering.tsx`
- `packages/components/src/sites-table/use-sites-table-p2-filtering.tsx` (for the mockup where P2s are in their own tab)
- `packages/components/src/sites-table/use-sites-table-plan-filtering.tsx`